extensions/csrc/cuda/rms_layernorm_kernel.cu

```diff
@@ -11,42 +11,33 @@
 #include "block_reduce.h"
 #include "../common/micros.h"
 #include "funcs/cast_functor.h"
-#include "funcs/op_functor.h"
+#include "funcs/binary_functor.h"
 
 using colossalAI::cuda::utils::block_reduce;
 using colossalAI::cuda::utils::ReduceType;
-using colossalAI::cuda::funcs::TypeConverter;
 using colossalAI::cuda::funcs::CastFunctor;
 using colossalAI::cuda::funcs::BinaryOpFunctor;
 using colossalAI::cuda::funcs::BinaryOpType;
 
-#define DISPATCH_RMSNORM_FLOAT_HALF_AND_BFLOAT(DATA_SIZE, TYPE, NAME, ...)  \
-  if (DATA_SIZE == 2) {                                                     \
-    switch (TYPE) {                                                         \
-      case at::ScalarType::Half: {                                          \
-        using scalar_t = at::Half;                                          \
-        __VA_ARGS__;                                                        \
-        break;                                                              \
-      }                                                                     \
-      case at::ScalarType::BFloat16: {                                      \
-        using scalar_t = at::BFloat16;                                      \
-        __VA_ARGS__;                                                        \
-        break;                                                              \
-      }                                                                     \
-      default:                                                              \
-        AT_ERROR(#NAME, " not implemented for '", toString(TYPE), "'");     \
-    }                                                                       \
-  } else {                                                                  \
-    switch (TYPE) {                                                         \
-      case at::ScalarType::Float: {                                         \
-        using scalar_t = float;                                             \
-        general_##__VA_ARGS__;                                              \
-        break;                                                              \
-      }                                                                     \
-      default:                                                              \
-        AT_ERROR(#NAME, " not implemented for '", toString(TYPE), "'");     \
-    }                                                                       \
-  }                                                                         \
+
+// Get type2 from type or vice versa (applied to half and bfloat16)
+template <typename T>
+struct TypeConverter {
+  using Type = half2;
+};
+
+#define TYPE_CONVERTER_SPECIALIZATION(FROM, TO)  \
```

Comment:
Wonder what are factors influenced decision to move `TypeConverter` at this location. It's only used by RMS layer norm kernel for now though.

Reply:
They are just used by RMSLayerNorm and it's not common, if it converts types between torch and Cuda built-in type, it's a good component, In fact, we need a component like this.
This part of code can be implemented by combine TypeConvert(Torch vs Cuda built-in type) with VecTypeTraits